### ISSUE TITLE
`heatmap` + `colormap=Makie.Categorical` results in incorrect `colorrange`

### ISSUE THREAD

**Issue by musoke** [MakieOrg/Makie.jl#3827]
- [x] are you running newest version (version from docs) ?
- [x] can you reproduce the bug with a fresh environment ? (`]activate --temp; add Makie`)
    - yes
- [x] What platform + GPU are you on?

    - GPU: NVIDIA RTX A3000
    -  `versioninfo()`:
```
Julia Version 1.10.3
Commit 0b4590a5507 (2024-04-30 10:59 UTC)
Build Info:
  Official https://julialang.org/ release
Platform Info:
  OS: Linux (x86_64-linux-gnu)
  CPU: 16 × 11th Gen Intel(R) Core(TM) i7-11850H @ 2.50GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-15.0.7 (ORCJIT, tigerlake)
Threads: 16 default, 0 interactive, 8 GC (on 16 virtual cores)
Environment:
  LD_LIBRARY_PATH = /snap/alacritty/70/usr/lib/x86_64-linux-gnu:/snap/alacritty/70/lib/x86_64-linux-gnu:/snap/alacritty/70/usr/lib/x86_64-linux-gnu/dri
  JULIA_NUM_THREADS = auto
```

I was attempting to use the [experimental categorical colorbars](https://docs.makie.org/stable/reference/blocks/colorbar/index.html#experimental_categorical_support) and found some incorrect behavior. 

I ran the following to set up a fresh environment
```
using Pkg
Pkg.activate(temp=true)
Pkg.add(["Makie", "CairoMakie"])
Pkg.status()
using CairoMakie
```
```
Status `/tmp/jl_c9NOuT/Project.toml`
  [13f3f980] CairoMakie v0.11.10
  [ee78f7c6] Makie v0.20.9
```

The following code without a categorical colorbar results in a figure as expected: 
```
fig, ax, pl = heatmap(1:3, 1:3, .-(1:3))
Colorbar(fig[1, 2], pl)
fig
```
![image](https://github.com/MakieOrg/Makie.jl/assets/16665084/66cd6cb0-1703-4dd7-9420-b5bef4637775)

However, using `colormap=Makie.Categorical` results in a colorbar whose `colorrange` does not correspond to the data range.
```
fig, ax, pl = heatmap(1:3, 1:3, .-(1:3), colormap=Makie.Categorical(:viridis))
Colorbar(fig[1, 2], pl)
fig
```
![image](https://github.com/MakieOrg/Makie.jl/assets/16665084/e1f5f1ab-5c14-4eb1-90a3-67bdc183c80e)
The colorbar is labeled from 1 to 4, none of which appear in the data.

**Comment by asinghvi17** [comment]
https://github.com/MakieOrg/Makie.jl/pull/3762 should fix this.

**Comment by musoke** [comment]
Cool!
I'll retry when 0.20.10 is released

**Comment by SimonDanisch** [comment]
On 0.20.10 this happens:
![image](https://github.com/MakieOrg/Makie.jl/assets/1010467/6b88e306-fc22-4b43-b120-a432db6bbe6b)

I think this is correct, since the transparent values are done by inserting NaNs.... 
It could be a bit confusing though since the user didn't input any NaNs. The only fix I can think of would be not creating the "sparse" heatmap from 3 vectors via NaNs but rather with a scatter plot or so... But that'd be a relatively big change

**Comment by musoke** [comment]
That colorbar is pretty close to what I would have expected.
I agree that this handling of "sparse" data is not be ideal long-term. Would it be possible to remove the NaN category from the colorbar after the fact?  Or internally use `missing` instead of `NaN` for such cases?  

It might be helpful for you to know more about my actual use case: I have an [`Enum`](https://docs.julialang.org/en/v1/base/base/#Base.Enums.@enum) and am plotting a heatmap where the colors correspond to enum instances.  I was originally doing this with a scatter plot but there are many points, so a heatmap is more readable.  I can writeup a MWE in another issue if that would be helpful.

**Comment by musoke** [comment]
How does colorrange work with categorical colorbars?  Especially in cases where the categories are not ordered or a category is not in the data but should appear on the axis.

**Comment by musoke** [comment]
I have tested against the since-released Makie v0.20.12 and v0.21.0 and get the output [promised above](https://github.com/MakieOrg/Makie.jl/issues/3827#issuecomment-2096597858).

Feel free to close this issue if the nans are better handled separately.
